File: template_project/python/04_validation_silver.py
# ============================================================
# PANEL ADMIN — ÉTAPE 4 : VALIDATION DE LA TABLE SILVER
# ============================================================
# Effectue les contrôles qualité sur la table Silver Iceberg.
#
# Contrôles :
#   1. Complétude des colonnes clés
#   2. Colonnes avec variations NA suspectes entre années
#   3. Doublons matricule × période
#   4. Cohérence des montants (net > brut)
#   5. Distribution des situations administratives
#
# Les rapports sont sauvegardés dans staging/panel_admin/validation/
#
# Architecture :
#   PySpark → lecture Iceberg (catalogue Nessie)
#   Polars  → tous les contrôles qualité
#
# Dépendances :
#   pip install polars boto3 python-dotenv pyspark
# ============================================================

import io
from dotenv import load_dotenv
import boto3
from botocore.client import Config
import polars as pl
from pyspark.sql import SparkSession

load_dotenv(".env")

# --- CONFIGURATION ---

# LORSQU'ON TRAVAILLE DEPUIS SA MACHINE LOCAL
MINIO_ENDPOINT   = "http://192.168.1.230:30137"
MINIO_ACCESS_KEY = "datalab-team"
MINIO_SECRET_KEY = "minio-datalabteam123"
NESSIE_URI       = "http://192.168.1.230:30604/api/v1"

# LORSQU'ON TRAVAILLE SUR JHUB
# MINIO_ENDPOINT   = "http://minio.mon-namespace.svc.cluster.local:80"
# MINIO_ACCESS_KEY = "datalab-team"
# MINIO_SECRET_KEY = "minio-datalabteam123"
# NESSIE_URI       = "http://nessie.trino.svc.cluster.local:19120/api/v1"

TABLE_SILVER = "nessie.silver.panel_admin_solde_mensuel"
BUCKET       = "staging"
PREFIX_VALID = "panel_admin/validation"

# --- CLIENT S3 (MinIO) ---
s3 = boto3.client(
    "s3",
    endpoint_url          = MINIO_ENDPOINT,
    aws_access_key_id     = MINIO_ACCESS_KEY,
    aws_secret_access_key = MINIO_SECRET_KEY,
    config                = Config(signature_version="s3v4"),
    region_name           = "us-east-1",
    verify                = False,
)


def sauvegarder_rapport(df, nom_fichier):
    csv_bytes = df.write_csv().encode("utf-8")
    s3.put_object(
        Bucket=BUCKET,
        Key=f"{PREFIX_VALID}/{nom_fichier}",
        Body=csv_bytes,
    )
    print(f"  ✓ Rapport : {nom_fichier}")


# ============================================================
# CONNEXION SPARK (lecture Iceberg uniquement)
# ============================================================

spark = (
    SparkSession.builder
    .appName("panel_admin_validation_silver")
    .config("spark.driver.memory", "16g")
    .config("spark.jars.packages",
            "org.apache.iceberg:iceberg-spark-runtime-3.5_2.12:1.6.1,"
            "org.apache.hadoop:hadoop-aws:3.3.4,"
            "org.projectnessie.nessie-integrations:nessie-spark-extensions-3.5_2.12:0.77.1")
    .config("spark.sql.extensions",
            "org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions,"
            "org.projectnessie.spark.extensions.NessieSparkSessionExtensions")
    .config("spark.sql.catalog.nessie",              "org.apache.iceberg.spark.SparkCatalog")
    .config("spark.sql.catalog.nessie.catalog-impl", "org.apache.iceberg.nessie.NessieCatalog")
    .config("spark.sql.catalog.nessie.uri",          NESSIE_URI)
    .config("spark.sql.catalog.nessie.ref",          "main")
    .config("spark.sql.catalog.nessie.warehouse",    "s3a://bronze/")
    .config("spark.hadoop.fs.s3a.endpoint",                   MINIO_ENDPOINT)
    .config("spark.hadoop.fs.s3a.access.key",                 MINIO_ACCESS_KEY)
    .config("spark.hadoop.fs.s3a.secret.key",                 MINIO_SECRET_KEY)
    .config("spark.hadoop.fs.s3a.path.style.access",          "true")
    .config("spark.hadoop.fs.s3a.impl",                       "org.apache.hadoop.fs.s3a.S3AFileSystem")
    .config("spark.hadoop.fs.s3a.aws.credentials.provider",
            "org.apache.hadoop.fs.s3a.SimpleAWSCredentialsProvider")
    .getOrCreate()
)

# ============================================================
# LECTURE SILVER → CONVERSION POLARS
# ============================================================

print("=" * 70)
print("POST-VALIDATION : CONTRÔLE QUALITÉ TABLE SILVER")
print("=" * 70)
print()

df = pl.from_pandas(spark.table(TABLE_SILVER).toPandas())
spark.stop()  # Spark n'est plus nécessaire après la lecture

nb_total = len(df)
nb_cols  = len(df.columns)
cols_all = df.columns

print(f"Table : {nb_total:,} lignes × {nb_cols} colonnes\n")

# ============================================================
# CONTRÔLE 1 : COMPLÉTUDE DES COLONNES CLÉS
# ============================================================

print("=" * 70)
print("CONTRÔLE 1 : COMPLÉTUDE DES COLONNES CLÉS")
print("=" * 70)
print()

colonnes_cles = [
    c for c in ["matricule", "nom", "montant_brut", "montant_net",
                "CODE_ORGANISME", "CODE_GRADE", "CODE_EMPLOI",
                "situation_normalisee", "periode"]
    if c in cols_all
]

for col in colonnes_cles:
    pct  = round(100.0 * df[col].is_not_null().sum() / nb_total, 1)
    flag = "⚠️ " if pct < 95 else "✓ "
    print(f"  {flag}{col:<30} : {pct:.1f}% complet")
print()

# ============================================================
# CONTRÔLE 2 : VARIATIONS NA PAR ANNÉE (colonnes suspectes)
# ============================================================

print("=" * 70)
print("CONTRÔLE 2 : COLONNES SUSPECTES (NA variable selon l'année)")
print("=" * 70)
print()

cols_a_surveiller = [
    c for c in ["matricule", "nom", "montant_brut", "montant_net",
                "organisme", "grade", "emploi", "lieu_affectation", "service"]
    if c in cols_all
]

if "periode" in cols_all:
    df_avec_annee = df.with_columns(
        pl.col("periode").str.slice(0, 4).alias("annee")
    )

    na_par_annee_rows = []
    for (annee,), groupe in df_avec_annee.group_by("annee"):
        row = {"annee": annee, "n": len(groupe)}
        for col in cols_a_surveiller:
            row[f"{col}_pct_na"] = round(100.0 * groupe[col].is_null().sum() / len(groupe), 1)
        na_par_annee_rows.append(row)

    na_par_annee = pl.DataFrame(na_par_annee_rows).sort("annee")

    colonnes_suspectes = []
    for col in cols_a_surveiller:
        col_na = f"{col}_pct_na"
        if col_na in na_par_annee.columns:
            pcts = na_par_annee[col_na].drop_nulls()
            if pcts.max() > 95 and pcts.min() < 5:
                colonnes_suspectes.append(col)
                print(f"⚠️  {col} : NA varie de {pcts.min():.1f}% à {pcts.max():.1f}% selon l'année")

    if not colonnes_suspectes:
        print("✓ Aucune colonne suspecte détectée")
    else:
        sauvegarder_rapport(na_par_annee, "na_par_annee_colonnes_surveillees.csv")
else:
    print("  (colonne 'periode' absente — contrôle ignoré)")
print()

# ============================================================
# CONTRÔLE 3 : DOUBLONS MATRICULE × PÉRIODE
# ============================================================

print("=" * 70)
print("CONTRÔLE 3 : DOUBLONS MATRICULE × PÉRIODE")
print("=" * 70)
print()

if "matricule" in cols_all and "periode" in cols_all:
    doublons = (
        df.filter(pl.col("matricule").is_not_null())
        .group_by(["matricule", "periode"])
        .agg(pl.len().alias("n"))
        .filter(pl.col("n") > 1)
        .sort("n", descending=True)
    )

    if len(doublons) > 0:
        print(f"⚠️  {len(doublons):,} combinaisons matricule × période en doublon")
        print("  Top 5 doublons :")
        for row in doublons.head(5).iter_rows(named=True):
            print(f"    {row['matricule']} | {row['periode']} | {row['n']} occurrences")
        sauvegarder_rapport(doublons, "doublons_matricule_periode.csv")
    else:
        print("✓ Aucun doublon matricule × période")
else:
    print("  (colonnes 'matricule' ou 'periode' absentes — contrôle ignoré)")
print()

# ============================================================
# CONTRÔLE 4 : COHÉRENCE MONTANTS (net > brut)
# ============================================================

print("=" * 70)
print("CONTRÔLE 4 : COHÉRENCE MONTANTS (net > brut)")
print("=" * 70)
print()

if "montant_net" in cols_all and "montant_brut" in cols_all:
    nb_inc = (
        df.filter(
            pl.col("montant_net").is_not_null()
            & pl.col("montant_brut").is_not_null()
            & (pl.col("montant_net") > pl.col("montant_brut"))
        )
        .select(pl.len())
        .item()
    )
    pct_inc = round(100.0 * nb_inc / nb_total, 2)
    if nb_inc > 0:
        print(f"⚠️  Net > Brut : {nb_inc:,} lignes ({pct_inc:.2f}%)")
    else:
        print("✓ Tous les montants sont cohérents (net ≤ brut)")
else:
    print("  (colonnes 'montant_net' ou 'montant_brut' absentes — contrôle ignoré)")
print()

# ============================================================
# CONTRÔLE 5 : DISTRIBUTION DES SITUATIONS
# ============================================================

print("=" * 70)
print("CONTRÔLE 5 : DISTRIBUTION DES SITUATIONS")
print("=" * 70)
print()

if "situation_normalisee" in cols_all:
    dist_sit = (
        df.group_by("situation_normalisee")
        .agg(pl.len().alias("nb"))
        .with_columns(
            (pl.col("nb") * 100.0 / nb_total).round(1).alias("pct")
        )
        .sort("nb", descending=True)
    )
    for row in dist_sit.iter_rows(named=True):
        print(f"  {row['situation_normalisee']:<25} : {row['nb']:,} ({row['pct']:.1f}%)")
else:
    print("  (colonne 'situation_normalisee' absente — contrôle ignoré)")
print()

# ============================================================
# RÉSUMÉ FINAL
# ============================================================

print("=" * 70)
print("RÉSUMÉ VALIDATION")
print("=" * 70)
print()

print(f"Lignes totales  : {nb_total:,}")
print(f"Colonnes        : {nb_cols}")

if "periode" in cols_all:
    print(f"Périodes        : {df['periode'].min()} à {df['periode'].max()} "
          f"({df['periode'].n_unique()} périodes)")

if "matricule" in cols_all:
    print(f"Agents uniques  : {df['matricule'].n_unique():,}")

print()
print("=" * 70)
print("✓ VALIDATION TERMINÉE")
print("=" * 70)
print("Prochaine étape : exécuter 05_silver_to_gold.py")

File: template_project/R/04_validation_silver.R
# ============================================================
# PANEL ADMIN — ÉTAPE 4 : VALIDATION DE LA TABLE SILVER
# ============================================================
# Adapté de : 03_post_validation_base_finale.R
#
# Effectue les mêmes contrôles qualité que le script original,
# mais directement sur la table Silver Iceberg via sparklyr/SQL
# (sans télécharger toute la table localement).
#
# Contrôles :
#   1. Colonnes avec variations NA suspectes entre années
#   2. Complétude des colonnes clés
#   3. Doublons matricule × période
#   4. Cohérence des montants (net > brut)
#   5. Distribution des situations administratives
#
# Les rapports sont sauvegardés dans staging/panel_admin/validation/
#
# Dépendances :
#   install.packages(c("sparklyr", "dplyr", "data.table", "aws.s3", "dotenv", "httr"))
# ============================================================

library(sparklyr)
library(dplyr)
library(data.table)
library(aws.s3)
library(dotenv)
library(httr)

httr::set_config(httr::config(ssl_verifypeer = FALSE, connecttimeout = 60, timeout = 600))

load_dot_env(file = ".env")

# --- CONFIGURATION ---

#LORSQU'ON TRAVAILLE DEPUIS SA MACHINE LOCAL
MINIO_ENDPOINT   <- "http://192.168.1.230:30137"
MINIO_ACCESS_KEY <- "datalab-team"
MINIO_SECRET_KEY <- "minio-datalabteam123"
NESSIE_URI       <- "http://192.168.1.230:30604/api/v1"

#LORSQU'ON TRAVAILLE SUR JHUB
# MINIO_ENDPOINT   <- "http://minio.mon-namespace.svc.cluster.local:80"
# MINIO_ACCESS_KEY <- "datalab-team"
# MINIO_SECRET_KEY <- "minio-datalabteam123"
# NESSIE_URI       <- "http://nessie.trino.svc.cluster.local:19120/api/v1"

#---------------------------------------------------------------------------------

TABLE_SILVER   <- "nessie.silver.panel_admin_solde_mensuel"
BUCKET         <- "staging"
PREFIX_VALID   <- "panel_admin/validation"
endpoint_propre <- sub("^https?://", "", MINIO_ENDPOINT)

Sys.setenv(
  AWS_ACCESS_KEY_ID     = MINIO_ACCESS_KEY,
  AWS_SECRET_ACCESS_KEY = MINIO_SECRET_KEY,
  AWS_DEFAULT_REGION    = "us-east-1"
)

sauvegarder_rapport <- function(df, nom_fichier) {
  tmp <- tempfile(fileext = ".csv")
  write.csv(df, tmp, row.names = FALSE)
  aws.s3::put_object(file = tmp, object = file.path(PREFIX_VALID, nom_fichier),
                     bucket = BUCKET, base_url = endpoint_propre, region = "",
                     use_https = FALSE)
  unlink(tmp)
  cat(sprintf("  ✓ Rapport : %s\n", nom_fichier))
}

# ============================================================
# CONNEXION SPARK
# ============================================================

config <- spark_config()
config$spark.driver.memory <- "16g"
config$spark.jars.packages <- paste(
  "org.apache.iceberg:iceberg-spark-runtime-3.5_2.12:1.6.1",
  "org.apache.hadoop:hadoop-aws:3.3.4",
  "org.projectnessie.nessie-integrations:nessie-spark-extensions-3.5_2.12:0.77.1",
  sep = ","
)
config$spark.sql.extensions <- paste(
  "org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions",
  "org.projectnessie.spark.extensions.NessieSparkSessionExtensions",
  sep = ","
)
config$spark.sql.catalog.nessie                  <- "org.apache.iceberg.spark.SparkCatalog"
config$`spark.sql.catalog.nessie.catalog-impl`   <- "org.apache.iceberg.nessie.NessieCatalog"
config$spark.sql.catalog.nessie.uri              <- NESSIE_URI
config$spark.sql.catalog.nessie.ref              <- "main"
config$spark.sql.catalog.nessie.warehouse        <- "s3a://bronze/"
config$spark.hadoop.fs.s3a.endpoint                    <- MINIO_ENDPOINT
config$spark.hadoop.fs.s3a.access.key                  <- MINIO_ACCESS_KEY
config$spark.hadoop.fs.s3a.secret.key                  <- MINIO_SECRET_KEY
config$spark.hadoop.fs.s3a.path.style.access           <- "true"
config$spark.hadoop.fs.s3a.impl                        <- "org.apache.hadoop.fs.s3a.S3AFileSystem"
config$`spark.hadoop.fs.s3a.aws.credentials.provider`  <- "org.apache.hadoop.fs.s3a.SimpleAWSCredentialsProvider"

sc <- spark_connect(master = "local", config = config)

cat(paste(rep("=", 70), collapse = ""), "\n")
cat("POST-VALIDATION : CONTRÔLE QUALITÉ TABLE SILVER\n")
cat(paste(rep("=", 70), collapse = ""), "\n\n")

df <- spark_read_table(sc, TABLE_SILVER)
nb_total <- sdf_nrow(df)
nb_cols  <- ncol(df)

cat(sprintf("Table : %s lignes × %d colonnes\n\n",
            format(nb_total, big.mark = ","), nb_cols))

# ============================================================
# CONTRÔLE 1 : COMPLÉTUDE DES COLONNES CLÉS
# ============================================================

cat(paste(rep("=", 70), collapse = ""), "\n")
cat("CONTRÔLE 1 : COMPLÉTUDE DES COLONNES CLÉS\n")
cat(paste(rep("=", 70), collapse = ""), "\n\n")

colonnes_cles <- c("matricule", "nom", "montant_brut", "montant_net",
                   "CODE_ORGANISME", "CODE_GRADE", "CODE_EMPLOI",
                   "situation_normalisee", "periode")
colonnes_cles <- intersect(colonnes_cles, colnames(df))

completude <- DBI::dbGetQuery(sc, sprintf("
  SELECT
    %s
  FROM %s
",
  paste(sprintf("ROUND(100.0 * COUNT(%s) / COUNT(*), 1) AS %s_pct",
                colonnes_cles, colonnes_cles), collapse = ",\n    "),
  TABLE_SILVER
))

for (col in colonnes_cles) {
  pct <- completude[[paste0(col, "_pct")]]
  flag <- if (pct < 95) "⚠️ " else "✓ "
  cat(sprintf("  %s%-30s : %.1f%% complet\n", flag, col, pct))
}
cat("\n")

# ============================================================
# CONTRÔLE 2 : VARIATIONS NA PAR ANNÉE (colonnes suspectes)
# ============================================================

cat(paste(rep("=", 70), collapse = ""), "\n")
cat("CONTRÔLE 2 : COLONNES SUSPECTES (NA variable selon l'année)\n")
cat(paste(rep("=", 70), collapse = ""), "\n\n")

# Calculer % NA par année pour les colonnes clés via SQL
cols_a_surveiller <- intersect(
  c("matricule", "nom", "montant_brut", "montant_net", "organisme",
    "grade", "emploi", "lieu_affectation", "service"),
  colnames(df)
)

na_par_annee <- DBI::dbGetQuery(sc, sprintf("
  SELECT
    SUBSTR(periode, 1, 4) AS annee,
    COUNT(*) AS n,
    %s
  FROM %s
  GROUP BY SUBSTR(periode, 1, 4)
  ORDER BY annee
",
  paste(sprintf("ROUND(100.0 * SUM(CASE WHEN %s IS NULL THEN 1 ELSE 0 END) / COUNT(*), 1) AS %s_pct_na",
                cols_a_surveiller, cols_a_surveiller), collapse = ",\n    "),
  TABLE_SILVER
))

# Identifier les colonnes avec forte variation
colonnes_suspectes <- character(0)
for (col in cols_a_surveiller) {
  col_na <- paste0(col, "_pct_na")
  if (col_na %in% names(na_par_annee)) {
    pcts <- na_par_annee[[col_na]]
    if (max(pcts, na.rm = TRUE) > 95 && min(pcts, na.rm = TRUE) < 5) {
      colonnes_suspectes <- c(colonnes_suspectes, col)
      cat(sprintf("⚠️  %s : NA varie de %.1f%% à %.1f%% selon l'année\n",
                  col, min(pcts, na.rm = TRUE), max(pcts, na.rm = TRUE)))
    }
  }
}

if (length(colonnes_suspectes) == 0) {
  cat("✓ Aucune colonne suspecte détectée\n")
} else {
  sauvegarder_rapport(na_par_annee, "na_par_annee_colonnes_surveillees.csv")
}
cat("\n")

# ============================================================
# CONTRÔLE 3 : DOUBLONS MATRICULE × PÉRIODE
# ============================================================

cat(paste(rep("=", 70), collapse = ""), "\n")
cat("CONTRÔLE 3 : DOUBLONS MATRICULE × PÉRIODE\n")
cat(paste(rep("=", 70), collapse = ""), "\n\n")

doublons <- DBI::dbGetQuery(sc, sprintf("
  SELECT matricule, periode, COUNT(*) AS n
  FROM %s
  WHERE matricule IS NOT NULL
  GROUP BY matricule, periode
  HAVING COUNT(*) > 1
  ORDER BY n DESC
", TABLE_SILVER))

if (nrow(doublons) > 0) {
  cat(sprintf("⚠️  %s combinaisons matricule × période en doublon\n",
              format(nrow(doublons), big.mark = ",")))
  cat("  Top 5 doublons :\n")
  for (i in 1:min(5, nrow(doublons))) {
    cat(sprintf("    %s | %s | %d occurrences\n",
                doublons$matricule[i], doublons$periode[i], doublons$n[i]))
  }
  sauvegarder_rapport(doublons, "doublons_matricule_periode.csv")
} else {
  cat("✓ Aucun doublon matricule × période\n")
}
cat("\n")

# ============================================================
# CONTRÔLE 4 : COHÉRENCE MONTANTS (net > brut)
# ============================================================

cat(paste(rep("=", 70), collapse = ""), "\n")
cat("CONTRÔLE 4 : COHÉRENCE MONTANTS (net > brut)\n")
cat(paste(rep("=", 70), collapse = ""), "\n\n")

incoherences <- DBI::dbGetQuery(sc, sprintf("
  SELECT COUNT(*) AS nb,
         ROUND(100.0 * COUNT(*) / %d, 2) AS pct
  FROM %s
  WHERE montant_net IS NOT NULL
    AND montant_brut IS NOT NULL
    AND montant_net > montant_brut
", nb_total, TABLE_SILVER))

if (incoherences$nb > 0) {
  cat(sprintf("⚠️  Net > Brut : %s lignes (%.2f%%)\n",
              format(incoherences$nb, big.mark = ","), incoherences$pct))
} else {
  cat("✓ Tous les montants sont cohérents (net ≤ brut)\n")
}
cat("\n")

# ============================================================
# CONTRÔLE 5 : DISTRIBUTION DES SITUATIONS
# ============================================================

cat(paste(rep("=", 70), collapse = ""), "\n")
cat("CONTRÔLE 5 : DISTRIBUTION DES SITUATIONS\n")
cat(paste(rep("=", 70), collapse = ""), "\n\n")

dist_sit <- DBI::dbGetQuery(sc, sprintf("
  SELECT situation_normalisee,
         COUNT(*) AS nb,
         ROUND(100.0 * COUNT(*) / %d, 1) AS pct
  FROM %s
  GROUP BY situation_normalisee
  ORDER BY nb DESC
", nb_total, TABLE_SILVER))

for (i in 1:nrow(dist_sit)) {
  cat(sprintf("  %-25s : %s (%.1f%%)\n",
              dist_sit$situation_normalisee[i],
              format(dist_sit$nb[i], big.mark = ","),
              dist_sit$pct[i]))
}
cat("\n")

# ============================================================
# RÉSUMÉ FINAL
# ============================================================

cat(paste(rep("=", 70), collapse = ""), "\n")
cat("RÉSUMÉ VALIDATION\n")
cat(paste(rep("=", 70), collapse = ""), "\n\n")

periodes_stats <- DBI::dbGetQuery(sc, sprintf("
  SELECT MIN(periode) AS premiere, MAX(periode) AS derniere,
         COUNT(DISTINCT periode) AS nb_periodes,
         COUNT(DISTINCT matricule) AS agents_uniques
  FROM %s
", TABLE_SILVER))

cat(sprintf("Lignes totales  : %s\n", format(nb_total, big.mark = ",")))
cat(sprintf("Colonnes        : %d\n", nb_cols))
cat(sprintf("Périodes        : %s à %s (%d périodes)\n",
            periodes_stats$premiere, periodes_stats$derniere, periodes_stats$nb_periodes))
cat(sprintf("Agents uniques  : %s\n", format(periodes_stats$agents_uniques, big.mark = ",")))
cat("\n")
cat(paste(rep("=", 70), collapse = ""), "\n")
cat("✓ VALIDATION TERMINÉE\n")
cat(paste(rep("=", 70), collapse = ""), "\n")
cat("Prochaine étape : exécuter 05_silver_to_gold.R\n")

spark_disconnect(sc)
